src/Razor/Razor.Runtime/src/Runtime/RazorRazorBasePage.cs

```diff
@@ -0,0 +1,468 @@
+        /// completion returns an empty <see cref="HtmlString"/>.</returns>
+        /// <remarks>The value returned is a token value that allows FlushAsync to work directly in an HTML
+        /// section. However the value does not represent the rendered content.
+        /// </remarks>
+        public async Task<HtmlString> FlushAsync()
+        {
+            // Calls to Flush are allowed if the page does not specify a Layout or if it is executing a section in the
+            // Layout.
+            if (!IsLayoutBeingRendered && !string.IsNullOrEmpty(Layout))
+            {
+                // TODO var message = Resources.FormatLayoutCannotBeRendered(Path, nameof(FlushAsync));
+                var message = $"Format Layout Cannot Be Rendered ({Path}, {nameof(FlushAsync)})";
+                throw new InvalidOperationException(message);
+            }
+            await Output.FlushAsync();
+            return HtmlString.Empty;
+        }
+
+        private void WriteUnprefixedAttributeValue(object value, bool isLiteral)
+        {
+            var stringValue = value as string;
+
+            // The extra branching here is to ensure that we call the Write*To(string) overload where possible.
+            if (isLiteral && stringValue != null)
+            {
+                WriteLiteral(stringValue);
+            }
+            else if (isLiteral)
+            {
+                WriteLiteral(value);
+            }
+            else if (stringValue != null)
+            {
+                Write(stringValue);
+            }
+            else
+            {
+                Write(value);
+            }
+        }
+
+        private void WritePositionTaggedLiteral(string value, int position)
+        {
+            BeginContext(position, value.Length, isLiteral: true);
+            WriteLiteral(value);
+            EndContext();
+        }
+
+        public abstract void BeginContext(int position, int length, bool isLiteral);
+
+        public abstract void EndContext();
+
+        private bool IsBoolFalseOrNullValue(string prefix, object value)
+        {
+            return string.IsNullOrEmpty(prefix) &&
+                (value == null ||
+                (value is bool && !(bool)value));
+        }
+
+        private bool IsBoolTrueWithEmptyPrefixValue(string prefix, object value)
+        {
+            // If the value is just the bool 'true', use the attribute name as the value.
+            return string.IsNullOrEmpty(prefix) &&
+                (value is bool && (bool)value);
+        }
+
+        public abstract void EnsureRenderedBodyOrSections();
+
+        private struct AttributeInfo
+        {
+            public AttributeInfo(
+                string name,
+                string prefix,
+                int prefixOffset,
+                string suffix,
+                int suffixOffset,
+                int attributeValuesCount)
```

Comment:
All of these integers can be removed and simplified. I'm OK if you want to leave them here for now for us to clean up, it's our mess after all.